k4FWCore/components/Reader.cpp

```diff
@@ -0,0 +1,129 @@
+    try {
+      auto out = (*this)();
+
+      auto outColls        = std::get<0>(out);
+      auto outputLocations = std::get<1>(out);
+
+      // if (out.size() != m_outputs.size()) {
+      //   throw GaudiException("Error during transform: expected " + std::to_string(m_outputs.size()) +
+      //                            " containers, got " + std::to_string(out.size()) + " instead",
+      //                        this->name(), StatusCode::FAILURE);
+      // }
+      for (unsigned i = 0; i != outColls.size(); ++i) {
+        auto objectp = std::make_unique<AnyDataWrapper<Out>>(std::move(outColls[i]));
+        if (auto sc = m_dataSvc->registerObject(outputLocations[i], objectp.get()); sc.isFailure()) {
+        }
+        // The store has the ownership so we shouldn't delete the object
+        auto ptr = objectp.release();
+      }
+      return Gaudi::Functional::FilterDecision::PASSED;
+    } catch (GaudiException& e) {
+      (e.code() ? this->warning() : this->error()) << e.tag() << " : " << e.message() << endmsg;
+      return e.code();
+    }
+  }
+
+  virtual std::tuple<vector_of_<Out>, std::vector<std::string>> operator()() const = 0;
+
+private:
+  // if In is a pointer, it signals optional (as opposed to mandatory) input
+  // template <typename T>
+  // using InputHandle_t = InputHandle_t<Traits_, std::remove_pointer_t<T>>;
+  // Gaudi::Property<std::vector<DataObjID>> m_inputLocations; // TODO/FIXME: remove this duplication...
+  // TODO/FIXME: replace vector of DataObjID property + call-back with a
+  //             vector<handle> property ... as soon as declareProperty can deal with that.
+  ServiceHandle<IDataProviderSvc> m_dataSvc{this, "EventDataSvc", "EventDataSvc"};
+};
+
+class Reader final : public CollectionPusher {
+public:
+  Reader(const std::string& name, ISvcLocator* svcLoc) : CollectionPusher(name, svcLoc) {}
+
+  // Gaudi doesn't run the destructor of the Services so we have to
+  // manually ask for the reader to be deleted so it will call finish()
+  // See https://gitlab.cern.ch/gaudi/Gaudi/-/issues/169
+  ~Reader() { iosvc->deleteReader(); }
+
+  ServiceHandle<IIOSvc> iosvc{this, "IOSvc", "IOSvc"};
+
+  StatusCode initialize() override {
+    if (!iosvc.isValid()) {
+      error() << "Unable to locate IIOSvc interface" << endmsg;
+      return StatusCode::FAILURE;
+    }
+
+    return StatusCode::SUCCESS;
+  }
+
+  StatusCode finalize() override {
+    if (iosvc) {
+      iosvc->deleteReader();
+    }
+    return StatusCode::SUCCESS;
+  }
+
+  // The IOSvc takes care of reading and passing the data
+  // By convention the Frame is pushed to the store
+  // so that it's deleted at the right time
+  std::tuple<std::vector<std::shared_ptr<podio::CollectionBase>>, std::vector<std::string>> operator()()
+      const override {
+    auto val = iosvc->next();
+
+    auto eds   = eventSvc().as<IDataProviderSvc>();
+    auto frame = std::move(std::get<2>(val));
```

Comment:
```suggestion
    auto frame = std::move(std::get<podio::Frame>(val));
```
Is this possible in this context? It would improve readability quite a bit.

Reply:
Fixed. Re `std::shared_ptr` they are used as substitute for the DataWrappers that we have in k4FWCore. For the collections that come from a Frame they are not needed, but for the ones that are created, for example, by other algorithms the std::shared_ptr take care of deleting them when they are not used. I've hidden the shared_ptr as much as I could and now they are not visible from the user side. The alternatives would be using DataWrapper since they can have ownership or not (but then we may as well use shared_ptr for that) or pushing raw pointers to the store and then doing manual memory management since Gaudi would delete the pushed pointers but not the collections they point to.